File: leetcode_src/src/TreePostorderTraversal.java
import java.util.*;
public class TreePostorderTraversal {
    public List<Integer> postorderTraversal(TreeNode root)
    {
        Deque<TreeNode> s = new LinkedList<>();
        List<Integer> res = new ArrayList<>();

        TreeNode node = root;
        TreeNode prev = null;
        while(!s.isEmpty() || node != null) {
            while(node != null) {
                s.push(node);
                node = node.left;
            }
            node = s.pop();
            if(node.right == null || node.right == prev) {
                res.add(node.val);
                prev = node;
                node = null;
            }
            else {
                s.push(node);
                node = node.right;
            }
        }

        return res;
    }
    public static void main(String[] args)
    {
        TreePostorderTraversal sol = new TreePostorderTraversal();
        TreeNode tree1 = BinaryTreeUtil.generatePerfectTree(3);
        tree1.printByLayer();
        System.out.println(sol.postorderTraversal(tree1));
        System.out.println(sol.postorderTraversal(null));
    }
}

File: leetcode_src/bin/TreePostorderTraversal.py
from typing import *
from TreeInorderTraversal import TreeNode

class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        s1 = []
        s2 = []

        if root: s1.append(root)
        while s1:
            head = s1.pop()
            s2.append(head.val)
            if head.left: s1.append(head.left)
            if head.right: s1.append(head.right)

        s2.reverse()
        return s2

    def postorderTraversal2(self, root: Optional[TreeNode]) -> List[int]:
        stack = []
        res = []
        prev = None

        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if not root.right or prev == root.right:
                res.append(root.val)
                prev = root
                root = None
            else:
                stack.append(root)
                root = root.right

        return res